src/tests/e2etest/ZtsiTests.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using NUnit.Framework;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Azure.Devices;
+using System.Text.Json;
+using System;
+using System.Threading.Tasks;
+using System.Text.RegularExpressions;
+
+namespace e2etesting
+{
+    [TestFixture]
+    public class ZtsiTests : E2eTest
+    {   
+        string ComponentName = "Ztsi";
+        public partial class Ztsi
+        {
+            public int Enabled { get; set; }
+            public string ServiceUrl { get; set; }
+        }
+        public partial class DesiredZtsi
+        {
+            public bool DesiredEnabled { get; set; }
+            public string DesiredServiceUrl { get; set; }
+        }
+
+        public partial class ResponseCode
+        {
+            public int ac { get; set; }
+        }
+        private int m_twinTimeoutSeconds;
+
+        [SetUp]
+        public void TestSetUp()
+        {
+            m_twinTimeoutSeconds = base.twinTimeoutSeconds;
+            base.twinTimeoutSeconds = base.twinTimeoutSeconds * 2;
+        }
+
+        [TearDown]
+        public void TestTearDown()
+        {
+            base.twinTimeoutSeconds = m_twinTimeoutSeconds;
+        }
+
+        [Test]      
+        public void ZtsiTest_Get()
+        {
+            if ((GetTwin().ConnectionState == DeviceConnectionState.Disconnected) || (GetTwin().Status == DeviceStatus.Disabled))
+            {
+                Assert.Fail("Module is disconnected or is disabled");
+            }
+
+            Ztsi deserializedReportedObject = JsonSerializer.Deserialize<Ztsi>(GetTwin().Properties.Reported[ComponentName].ToString());
+            Assert.True(IsRegexMatch(new Regex(@"[0-2]"), deserializedReportedObject.Enabled));
+            Assert.True(deserializedReportedObject.ServiceUrl is string);
+        }
+
+        [Test]
+        public void ZtsiTest_Set_Get()
+        {
+            var desiredZtsi = new DesiredZtsi
+            {
+                DesiredEnabled = true,
+                DesiredServiceUrl = "https://ztsiendpoint.com"
+            };
+
+            var expectedZtsi = new Ztsi
+            {
+                Enabled = 1,
+                ServiceUrl = desiredZtsi.DesiredServiceUrl
+            };
+
+            if ((GetNewTwin().ConnectionState == DeviceConnectionState.Disconnected) || (GetTwin().Status == DeviceStatus.Disabled))
+            {
+                Assert.Fail("Module is disconnected or is disabled");
+            }
+
+            Ztsi deserializedReportedObject = JsonSerializer.Deserialize<Ztsi>(GetTwin().Properties.Reported[ComponentName].ToString());
+
+            Twin twinPatch = CreateTwinPatch(ComponentName, desiredZtsi);
+
+            if (UpdateTwinBlockUntilUpdate(twinPatch))
```

Comment:
For example, `Enabled = true` and `ServiceUrl = ""` is an invalid configuration. If the update to `Enabled` state is sent to OSConfig first it will fail and the second update to `ServiceUrl` will succeed, which explains why the `ServiceUrl` is updated but not `Enabled`.

Reply:
Suggest to break this into 2 separate updates:

1. First to `DesiredServiceUrl` (wait for successful update)
2. Then to `DesiredEnabled`

This should ensure the correct ordering and guarantee valid configurations

Reply:
Thanks Robert!